lib/Arch.cpp

```diff
@@ -0,0 +1,441 @@
+}  // namespace remill
+
+namespace anvill {
+
+// Try to recover parameter names using debug information. Otherwise, name the
+// parameters with the form "param_x". The mapping of the return value is
+// positional starting at 1.
+std::map<unsigned, std::string> TryRecoverParamNames(
+    const llvm::Function &function) {
+  std::map<unsigned int, std::string> param_names;
+
+  for (auto &block : function) {
+    for (auto &inst : block) {
+      if (auto debug_inst = llvm::dyn_cast<llvm::DbgInfoIntrinsic>(&inst)) {
+        if (auto value_intrin = llvm::dyn_cast<llvm::DbgDeclareInst>(&inst)) {
+          const llvm::MDNode *mdn = value_intrin->getVariable();
+          const llvm::DILocalVariable *div =
+              llvm::cast<llvm::DILocalVariable>(mdn);
+
+          // Make sure it is actually an argument
+          if (div->getArg() != 0) {
+            LOG(INFO) << div->getArg() << " : " << div->getName().data();
+            param_names[div->getArg()] = div->getName().data();
+          }
+        } else if (auto value_intrin =
+                       llvm::dyn_cast<llvm::DbgValueInst>(debug_inst)) {
+          const llvm::MDNode *mdn = value_intrin->getVariable();
+          const llvm::DILocalVariable *div =
+              llvm::cast<llvm::DILocalVariable>(mdn);
+
+          if (div->getArg() != 0) {
+            LOG(INFO) << div->getArg() << " : " << div->getName().data();
+            param_names[div->getArg()] = div->getName().data();
+          }
+        }
+      }
+    }
+  }
+
+  // If we don't have names for some parameters then automatically name them
+  unsigned int num_args =
+      (unsigned int)(function.args().end() - function.args().begin());
+  for (unsigned int i = 1; i <= num_args; i++) {
+    if (!param_names.count(i)) {
+      param_names[i] = "param" + std::to_string(i);
+    }
+  }
+
+  return param_names;
+}
+
+// TODO: This doesn't really fit here so I will need to find a better place for
+// this. Translates an llvm::Type to a type that conforms to the spec in
+// TypeParser.cpp
+std::string TranslateType(const llvm::Type &type) {
+  unsigned int id = type.getTypeID();
+
+  std::string ret;
+
+  // Types should match the type parsing in TypeParser.cpp
+  switch (id) {
+    case llvm::Type::VoidTyID: {
+      ret = "v";
+      break;
+    }
+    case llvm::Type::HalfTyID: {
+      ret = "e";
+      break;
+    }
+    case llvm::Type::FloatTyID: {
+      ret = "f";
+      break;
+    }
+    case llvm::Type::DoubleTyID: {
+      ret = "d";
+      break;
+    }
+    case llvm::Type::X86_FP80TyID: {
+      ret = "D";
+      break;
+    }
+    case llvm::Type::X86_MMXTyID: {
+      ret = "M";
+      break;
+    }
+    case llvm::Type::IntegerTyID: {
+      auto derived = llvm::cast<llvm::IntegerType>(type);
+      // Since there is no way to check for uint vs. int, lower all integer
+      // types to signed integers. Maybe come back when there is a better way to
+      // figure out unsigned vs signed ints
```

Comment:
Could look for usages, e.g. is it used in an `add`, and if so, is it signed or unsigned overflow attributed, or `div` vs. `sdiv` kind of thing. File an issue for it, add a TODO comment referencing the issue.